### DISCUSSION TITLE
Allow limiting the maximum volume of a media player

### DISCUSSION THREAD

**Discussion by emontnemery** [home-assistant/architecture#984] [Entity Models]
## Proposal

Allow limiting the maximum volume of a media player to protect speakers etc.

### Complications

Media player services `volume_up` + `volume_down` increase and decrease the media player's volume by an undetermined amount. As such they are incompatible with a volume limit.

Possible ways forward could be to change media player services `volume_up` + `volume_down` to either of:
1. Never support `media_player.volume_up` + `media_player.volume_down` for media players which support `media_player.volume_set`
2. Make the base class implement `media_player.volume_up` + `media_player.volume_down` for media players which support `media_player.volume_set`

**Comment by elupus** [comment]
I'm mostly in agreement.

There is case we currently don't handle to well with this proposal. It's hammering on volume up. Since we expect volume state to reflect the current device state. Update of this state can be delayed from a set volume command. We would have to maintain some memory of the modified volume, until an actual state result comes in. Probably even two in succession with same volume, since when state is reported back, another request may be queued up.

Note. This is not specific to volume, it's specific to anything where we implement a step wise change with a absolute set of value. (fan comes to mind since i have issus with this).

Reason for queued up step wise request could stem from for example a rotary controller wich just triggers up/down in rapid succession.

**Reply by emontnemery** [reply]
Interesting. Would it be easier if the memory is of previous calls to the step wise service though?
That is, the first time you call "volume_up", volume is increased by 10, the next time by 20 and so on, to be reset to 10 when there's a state update from the device?

**Reply by elupus** [reply]
Yes a memory of calls to service could work too with some base point. But you would have to handle up,up,up,down,up, and problem with delayed response remain.

up,up,up,state update from two ups, up , state update from three ups, state update from four ups.

So you cant really reset the base point on first state update, since you would miss one up. State have to have stabilized.

**Comment by mvdwetering** [comment]
>Possible ways forward could be to change media player services volume_up + volume_down to either of:
>
>1. Never support media_player.volume_up + media_player.volume_down for media players which support media_player.volume_set
>1. Make the base class implement media_player.volume_up + media_player.volume_down for media players which support media_player.volume_set

I see that `volume_up` could go over a configured limit, but removing it seems a bit too much. 

1. Volume step commands are very useful for receiver/tv mediaplayers which, most of the time, can also set volume. 
Multiple receiver and TV integrations use the native "volume step" commands offered by the APIs on the receivers/tvs for `volume_up`/`volume_down` services. These native step commands usually have increments the same as pressing the volume up/down buttons on the remote (or an app with volume up/down feature) would do. 
When wanting to control volume in steps from automations or the frontend the `volume_up`/`volume_down` services are an easy way to get the same stepsize as the normal remote/app. When those services are not available one would need to calculate the size of those steps and adding/subtracting them and then call `volume_set` which is a hassle to do in automations and an easy source of errors.


2. I am not sure how implementing volume_up/volume_down in the baseclass would be different from the current default implementation for volume_up/volume_down that uses volume steps of 0.1 (10%). Those steps are usually way too big for devices like receivers and tvs (but might be correct for other mediaplayers). Currently that is not a problem as the default implementation is overridden by one that uses the native step commands. Having one fixed default step size would be a problem.
As an example Yamaha receivers typically have a volume range of -80.5 to 16.5dB with stepsize 0.5dB so a step in HA volume range is about 0,00515. Most tvs I know seem to have a volume range of 0-100 and volume step size of 1 , so a step of  0.01 in HA volume range.  
With these different "optimal" stepsizes, the integration should probably be able to provide it.

edit: some markup fixes and minor rewording
